File: java/src/chapter2/section1/E18_Tracing.java
package chapter2.section1;

import java.util.HashSet;
import java.util.Set;

import edu.princeton.cs.algs4.Draw;
import utils.Utils;

public class E18_Tracing {
  private static int comparisons = 0;
  private static int swaps = 0;

  public static void main(String[] args) {
    if (args.length < 2) {
      System.err.println("Usage: java Tracing <algorithm> <array size>");
      System.exit(1);
    }

    String algorithm = args[0];
    int n = Integer.parseInt(args[1]);

    Double[] A = Utils.generateRandomDoubleArray(n);

    // Create a Draw object
    Draw draw = new Draw();

    // Set canvas size
    draw.setCanvasSize(1500, 300);

    // Set scale with padding
    draw.setXscale(-2, A.length + 2); // Adjust x-scale for left and right padding
    draw.setYscale(-0.05, 1.15); // Adjust y-scale for less bottom padding

    if (algorithm.equalsIgnoreCase("Selection")) {
      selectionSort(draw, A);
    } else if (algorithm.equalsIgnoreCase("Insertion")) {
      insertionSort(draw, A);
    } else if (algorithm.equalsIgnoreCase("Shell")) {
      shellSort(draw, A);
    } else {
      System.err.println("Unknown algorithm: " + algorithm);
      System.exit(1);
    }

    rerender(draw, A, -1, -1, new HashSet<Integer>(), algorithm, 0);
  }

  private static <T extends Comparable<T>> void rerender(Draw draw, T[] A, int iteration, int destination,
      Set<Integer> traceSet, String algorithm, int h) {
    // Clear the canvas
    draw.clear(new java.awt.Color(239, 224, 185)); // Light coffee color background

    // Draw the algorithm name, current iteration, number of comparisons, and swaps
    draw.setPenColor(Draw.BLACK);
    String title = algorithm + "Sort";
    if (iteration >= 0) {
      title += " | i: " + iteration;
    }
    title += " | Comp.: " + comparisons + " | Swaps: " + swaps;
    if (algorithm == "Shell")
      title += " | h: " + h;
    draw.text(A.length / 2.0, 1.05, title);

    // Draw the background grid lines
    draw.setPenColor(new java.awt.Color(211, 211, 211)); // Light grey for grid lines
    draw.setPenRadius(0.001);
    for (double i = 0; i <= 1.0; i += 0.1) {
      draw.line(-2, i, A.length + 2, i); // Horizontal grid lines with padding
    }
    for (int i = 0; i <= A.length; i++) {
      draw.line(i, 0, i, 1.0); // Vertical grid lines
    }

    // Draw the axes and scale lines
    draw.setPenColor(Draw.BLACK);
    draw.setPenRadius(0.002);

    // Draw bars
    for (int i = 0; i < A.length; i++) {
      double x = i + 0.5;
      double y = ((Double) A[i]) / 2.0;
      double halfWidth = 0.4;
      double halfHeight = ((Double) A[i]) / 2.0;

      if (i == destination) {
        draw.setPenColor(231, 76, 60); // Red modern
      } else if (traceSet.contains(i)) {
        draw.setPenColor(44, 62, 80); // Black modern
      } else {
        draw.setPenColor(149, 165, 166); // Gray modern
      }

      draw.filledRectangle(x, y, halfWidth, halfHeight);
    }
    draw.enableDoubleBuffering();
    draw.show();
    // Add a slight delay to visualize the sorting process
    try {
      Thread.sleep(250); // 200 milliseconds delay
    } catch (InterruptedException e) {
      e.printStackTrace();
    }
  }

  private static <T extends Comparable<T>> boolean less(T v, T w) {
    comparisons++;
    boolean result = v.compareTo(w) < 0;
    return result;
  }

  private static <T extends Comparable<T>> void exch(T[] A, int i, int j) {
    swaps++;
    T temp = A[i];
    A[i] = A[j];
    A[j] = temp;
  }

  public static <T extends Comparable<T>> void selectionSort(Draw draw, T[] A) {
    int N = A.length;
    for (int i = 0; i < N - 1; i++) {
      Set<Integer> traceSet = new HashSet();
      int min = i;
      for (int j = i + 1; j < N; j++) {
        if (less(A[j], A[min])) {
          min = j;
        }
        traceSet.add(j);
      }
      exch(A, i, min);
      rerender(draw, A, i, i, traceSet, "Selection", 0); // Highlight swapped bars
    }
  }

  public static <T extends Comparable<T>> void insertionSort(Draw draw, T[] A) {
    int N = A.length;
    for (int i = 1; i < N; i++) {
      Set<Integer> traceSet = new HashSet();
      int min = -1;
      for (int j = i; j > 0 && less(A[j], A[j - 1]); j--) {
        exch(A, j, j - 1);
        min = j - 1;
        traceSet.add(j);
        traceSet.add(j - 1);
      }
      rerender(draw, A, i, min, traceSet, "Insertion", 0); // Highlight swapped bars
    }
  }

  public static <T extends Comparable<T>> void shellSort(Draw draw, T[] A) {
    int N = A.length;
    int h = 1;
    while (h < N / 3)
      h = h * 3 + 1;

    while (h >= 1) {
      for (int i = h; i < N; i++) {
        Set<Integer> traceSet = new HashSet();
        int min = -1;
        for (int j = i; j >= h && less(A[j], A[j - h]); j -= h) {
          exch(A, j, j - h);
          min = j - h;
          traceSet.add(j);
          traceSet.add(j - h);
        }
        rerender(draw, A, i, min, traceSet, "Shell", h); // Highlight swapped bars
      }
      h /= 3;
    }
  }
}

File: python/chapter2/section1/E18_Tracing.py
import sys
import random
import matplotlib.pyplot as plt

plt.rcParams['font.family'] = 'Arial'

comparisons = 0
swaps = 0
ligth_coffee_color = (239/255, 224/255, 185/255)
red_color = (255/255, 69/255, 0/255)
green_color = (34/255, 139/255, 34/255)

red_modern_color = (231/255, 76/255, 60/255)
black_modern_color = (44/255, 62/255, 80/255)
gray_modern_color = (149/255, 165/255, 166/255)
green_emerald_color = (80/255, 200/255, 120/255)
speed_delay = 0.5

blue_modern_color = (0/255, 120/255, 210/255)
white_modern_color = (220/255, 220/255, 220/255)
green_modern_color = (0/255, 158/255, 96/255)

def generate_random_array(n):
    return [random.random() for _ in range(n)]

def less(v, w):
    global comparisons
    comparisons += 1
    return v < w

def exch(A, i, j):
    global swaps
    swaps += 1
    A[i], A[j] = A[j], A[i]

def rerender(A, destination, traceSet = set(), algorithm = "", h=0):
    plt.clf()
    fig = plt.gcf()

    fig.patch.set_facecolor(ligth_coffee_color)

    ax = plt.gca()
    ax.set_facecolor(ligth_coffee_color)

    # Hide all spines except the bottom one
    ax.spines['top'].set_visible(False)
    ax.spines['right'].set_visible(False)
    ax.spines['left'].set_visible(False)
    ax.spines['bottom'].set_position('zero')  # Position the bottom spine (x-axis) at y=0

    ax.set_xlim(-0.5, len(A) - 0.5)
    
    # Set the x-axis limits and display only the first and last values along with some intermediate values
    ax.set_xticks([0] + list(range(1, len(A)-1, max(1, len(A)//10))) + [len(A)-1])
    ax.set_xticklabels([str(x) for x in ax.get_xticks()], fontfamily='Arial')
    ax.xaxis.set_tick_params(size=0)  # Remove the tick marks

    ax.yaxis.set_visible(False)  # Hide the y-axis

    plt.subplots_adjust(left=0.01, right=0.99, top=0.85, bottom=0.13)  # Minimize margins

    plt.bar(range(len(A)), A, width=0.95, color=[
        red_modern_color if i == destination else
        black_modern_color if i in traceSet else
        gray_modern_color for i in range(len(A))
    ])
    title = f"{algorithm} Sort"

    subtitle = f"Comp.: {comparisons}, Swaps: {swaps}"
    if algorithm == "Shell":
        subtitle += f", h: {h}"

    plt.title(title, fontsize=18, loc="left", fontfamily='Arial', fontweight='bold')
    plt.text(-0.5, 1.01, subtitle, fontsize=14, fontfamily='Arial')
    
    plt.text(-0.5, -0.2, "Source: Farrukh Normuradov", ha='left', fontsize=10, color='gray', fontfamily='Arial')
    plt.ylim(-0.05, 1.10)

    ax.add_patch(plt.Rectangle((0.01,0.98),             # Set location of rectangle by lower left corder
                            0.05,                       # Width of rectangle
                            0.55,                       # Height of rectangle.
                            facecolor=blue_modern_color, 
                            transform=fig.transFigure, 
                            clip_on=False, 
                            linewidth = 0))
    
    ax.add_patch(plt.Rectangle((0.94, 0.02),             # Set location of rectangle by lower left corder
                            0.05,                       # Width of rectangle
                            -0.55,                       # Height of rectangle.
                            facecolor=green_modern_color, 
                            transform=fig.transFigure, 
                            clip_on=False, 
                            linewidth = 0))

    # Add delay based on speed
    try:
        global speed_delay
        plt.pause(speed_delay)
    except Exception as e:
        print(e)

def selection_sort(A):
    N = len(A)
    for i in range(N - 1):
        traceSet = set()
        min_idx = i
        for j in range(i + 1, N):
            traceSet.add(j)
            if less(A[j], A[min_idx]):
                min_idx = j
        rerender(A, min_idx, traceSet, "Selection")
        exch(A, i, min_idx)

def insertion_sort(A):
    N = len(A)
    for i in range(1, N):
        min_idx = -1
        traceSet = set()
        for j in range(i, 0, -1):
            traceSet.add(j)
            traceSet.add(j-1)
            if less(A[j], A[j - 1]):
                exch(A, j, j - 1)
                min_idx = j - 1
            else:
                break
        rerender(A, min_idx, traceSet, "Insertion")

def shell_sort(A):
    N = len(A)
    h = 1
    while h < N // 3:
        h = h * 3 + 1
    while h >= 1:
        for i in range(h, N):
            traceSet = set()
            min_idx = -1
            for j in range(i, h-1, -h):
                traceSet.add(j)
                traceSet.add(j-h)
                if less(A[j], A[j - h]):
                    exch(A, j, j - h)
                    min_idx = j - h
                else:
                    break
            rerender(A, min_idx, traceSet, "Shell", h)
        h //= 3

def main():
    if len(sys.argv) < 3:
        print("Usage: python Animation.py <algorithm> <array size> <speed>")
        sys.exit(1)

    algorithm = sys.argv[1]
    n = int(sys.argv[2])
    A = generate_random_array(n)
    speed = sys.argv[3].lower() if len(sys.argv) > 3 else 'normal'

    # Set speed delay based on input
    global speed_delay
    if speed == 'slow':
        speed_delay = 1.0
    elif speed == 'normal':
        speed_delay = 0.5
    elif speed == 'fast':
        speed_delay = 0.01
    else:
        print("Unknown speed: ", speed)
        sys.exit(1)

    plt.ion()
    fig = plt.figure(figsize=(15, 3))
    fig.patch.set_facecolor(ligth_coffee_color)  # Set the background color to light coffee

    if algorithm.lower() == "selection":
        selection_sort(A)
    elif algorithm.lower() == "insertion":
        insertion_sort(A)
    elif algorithm.lower() == "shell":
        shell_sort(A)
    else:
        print("Unknown algorithm:", algorithm)
        sys.exit(1)

    rerender(A, -1, algorithm=algorithm)

    plt.ioff()
    plt.show()

if __name__ == "__main__":
    main()